### DISCUSSION TITLE
Potential bug in ParameterDict get from shared dict

### DISCUSSION THREAD

**Discussion by meissnereric** [apache/incubator-mxnet#13919] [Q&A]
## Description

The Gluon Block's collect_param method does not surface the parameters of "shared" ParameterDicts and we need this functionality for some other work. 

I think this is a bug in the implementation of the ParameterDict class's [__getitem__ functionality](https://mxnet.incubator.apache.org/_modules/mxnet/gluon/parameter.html), but I definitely could be missing something. The ```get``` method looks in the shared dict, but the ```___getitem___``` method doesn't, and the ```___getitem___``` is what the Gluon Block's ```collect_param```s method uses. I don't need it to add items to the shared dict, just for them to show up when I loop through the overall ParameterDict.

## Environment info (Required)

```
----------Python Info----------
Version      : 3.6.0
Compiler     : GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.38)
Build        : ('default', 'Nov  8 2017 15:43:58')
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 18.0
Directory    : /Users/erimeiss/.pyenv/versions/3.6.0/lib/python3.6/site-packages/pip
----------MXNet Info-----------
/Users/erimeiss/.pyenv/versions/3.6.0/lib/python3.6/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Version      : 1.3.0
Directory    : /Users/erimeiss/.pyenv/versions/3.6.0/lib/python3.6/site-packages/mxnet
Commit Hash   : b3be92f4a48bce62a5a8424271871c2f81c8f7f1
----------System Info----------
Platform     : Darwin-16.7.0-x86_64-i386-64bit
system       : Darwin
node         : 8c8590165f92.ant.amazon.com
release      : 16.7.0
version      : Darwin Kernel Version 16.7.0: Sun Oct 28 22:30:19 PDT 2018; root:xnu-3789.73.27~1/RELEASE_X86_64
----------Hardware Info----------
machine      : x86_64
processor    : i386
b'machdep.cpu.extfeatures: SYSCALL XD 1GBPAGE EM64T LAHF LZCNT PREFETCHW RDTSCP TSCI'
b'machdep.cpu.leaf7_features: SMEP ERMS RDWRFSGS TSC_THREAD_OFFSET BMI1 HLE AVX2 BMI2 INVPCID RTM SMAP RDSEED ADX IPT SGX FPU_CSDS MPX CLFSOPT'
b'machdep.cpu.features: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX SMX EST TM2 SSSE3 FMA CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC MOVBE POPCNT AES PCID XSAVE OSXSAVE SEGLIM64 TSCTMR AVX1.0 RDRAND F16C'
b'machdep.cpu.brand_string: Intel(R) Core(TM) i7-7660U CPU @ 2.50GHz'
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0196 sec, LOAD: 0.5705 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.0288 sec, LOAD: 0.8377 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.0347 sec, LOAD: 0.7449 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0511 sec, LOAD: 1.0068 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0164 sec, LOAD: 0.8213 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0266 sec, LOAD: 0.4599 sec.
```

## Minimum reproducible example
(If you are using your own code, please provide a short script that reproduces the error. Otherwise, please provide link to the existing example.)

```python
from mxnet.gluon import ParameterDict
outer_params = ParameterDict()
outer_params.get("a", shape=(2,2))
outer_params.initialize()

from mxnet.gluon import HybridBlock
class ExampleBlock(HybridBlock):
    def __init__(self, param_dict, prefix=''):
        super(ExampleBlock, self).__init__(prefix=prefix, params=param_dict)

eb = ExampleBlock(outer_params)


eb.collect_params()
```

Running this, I would expect it to output the ```a``` parameter, but instead it outputs just an empty parameter dictionary because collect_params doesn't look inside the shared param_dict.

## What have you tried to solve it?

To unblock temporarily, we're just using a workaround in our code and just manually overriding the ```_params``` of the Block with the correct ParameterDict, but that's not a real option even in the short term. The obvious solution I see would be to make``` __getitem__``` also loop over the shared ParameterDict's item's but I don't know what issues that might cause elsewhere.

[a comment by an automated account is not shown]

**Comment by piyushghai** [comment]
Hi @meissnereric Thanks for posting this. This seems like a question to me and this can get more attention on the MXNet discussion forum (https://discuss.mxnet.io), where it will get a wider audience and allow others to learn as well.
Thanks!

@mxnet-label-bot add [Question, Gluon]

**Comment by meissnereric** [comment]
Hey Piyush,

Sure I'm happy to post this on the forum as well as it probably does need some sort of discussion, but I do think its more than a misunderstanding so can we leave the issue open until the discussion resolves? 

[Discussion post](https://discuss.mxnet.io/t/potential-bug-in-parameterdict-get-from-shared-dict/2793)

Thanks,
Eric
